simply/actor.py

```diff
@@ -79,52 +80,406 @@ class Actor:
+                # time step will use the reduced demand for the time steps afterwards
+                soc_prediction = self.predict_socs(clip=True, clip_value=1)
+
+            # if the predicted soc is 1 for the time steps before the current one, it is not
+            # possible to buy energy before. Since planning used only for time step 0, rest can
+            # be skipped
+            if max(soc_prediction[:i+1]) >= 1-cfg.config.EPS:
+                # set planning horizon for soc_prediction
+                self.planning_horizon = i
+                break
+
+        soc_prediction = self.predict_socs(clip=False)
+        self.predicted_soc = soc_prediction
+
+        return self.market_schedule
+
+    def plan_selling_strategy(self):
+        """Returns market_schedule where overcharges are sold at the highest possible price.
+
+        This strategy extends strategy 1. If the soc would rise over 1 the
+        market_schedule will sell the energy by finding the most expensive time slot before or at
+        the energy need time slot. Soc constrains will be considered, meaning that if needed,
+        multiple time slots will be used to sell the energy.
+        :return: market_schedule
+        """
+        # set planning horizon for soc_prediction
+        self.planning_horizon = self.horizon-1
+        soc_prediction = self.predict_socs(clip=False)
+        soc_prediction = soc_prediction[:self.planning_horizon+1]
+
+        # ToDo selling is not capped by max c-rate of battery
+        # iterate over socs
+        for i, _ in enumerate(soc_prediction):
+            energy = soc_prediction[i] * self.battery.capacity
+            overcharge = energy - self.battery.capacity
+            # if overcharge is found, find the possible prices to sell this energy
+            while overcharge > 0:
+                possible_prices = self.pred.selling_price.copy()[:self.planning_horizon+1]
+                possible_prices[soc_prediction < 0 + cfg.config.EPS] = float('-inf')
+                # in between now and the peak, the right most/latest zero soc does not allow
+                # reducing the soc before. Energy most be sold afterwards
+                zero_soc_indices = np.where(soc_prediction[:i] < 0 + cfg.config.EPS)
+                if np.any(zero_soc_indices):
+                    right_most_peak = np.max(zero_soc_indices)
+                else:
+                    right_most_peak = 0
+                possible_prices[:right_most_peak] = float('-inf')
+                highest_price_index = np.argmax(possible_prices[:i + 1])
+
+                # if the energy is sold at the time of production no storage is needed
+                if highest_price_index == i:
+                    self.market_schedule[i] -= overcharge
+                    soc_prediction = self.predict_socs(clip=False)
+                    break
+
+                soc_to_zero = min(np.min(soc_prediction[highest_price_index:i + 1]), 1)
```

Comment:
Added comment, removed outer min() call since socs_to_zero above 1 should not be possible. Instead an assertion was added to make sure this assumption is correct.